### ISSUE TITLE
[Build] Links are not properly referenced in local build.

### ISSUE THREAD

**Issue by mdynnl** [symfony/symfony-docs#18187]
For example, at `/quick_tour/the_big_picture.html`, the link `Flex: Compose your Application` points to `/flex_recipes.html` instead of `/quick_tour/the_big_picture.html`.

I already tracked down the cause and can work on a PR but need further pointer as the issue cames from https://github.com/symfony-tools/docs-builder.

**Comment by xabbuh** [comment]
this seems to have been fixed in the meantime

**Comment by mdynnl** [comment]
Not sure if I'm missing something. I checked the commits in both repos and git blame of the code with the issue but can't find anyting related to the fix.

The issue rises from 1) the use of `<base href="/">` as a mean to fix the image paths and asset paths which was the original cause as 2) public image path is configured to use relative and 3) `asset` (from `docs-builder`) is also returning relative url instead of `/` absolute for some reason (which could still be fixed from here by overriding asset extension class)

**Comment by xabbuh** [comment]
Do you still experience the issue? I cannot see it.

**Comment by mdynnl** [comment]
Yes, I ran this
```bash
php build.php --disable-cache
php -S localhost:8000 -t output
```
and take the same steps as the first comment. Like it works from the index page, but the link at the bottom of `The Big Picture` to `Flex: Compose your Application` is broken.

**Comment by xabbuh** [comment]
That was my mistake. I thought you were talking about the documentation rendered on symfony.com.

**Comment by javiereguiluz** [comment]
This issue is unfortunate, but I'm afraid we don't have many resources to try to fix this.

Do broken links happen in the entire docs ... or only in the pages that belong to "The Quick Tour" section?

**Comment by mdynnl** [comment]
I've only tested said links but probably the entire site has this issue on local build. I'll try to find time to fix the issue this week.

**Comment by mdynnl** [comment]
Should've put this a month ago. Below are the diffs that should fix this issue. But I'm not sure if the changes to docs-builder would break the symfony docs.

https://github.com/symfony-tools/docs-builder/blob/faf2c5f9cd22dd00c8ef9614d521ea1cdff43390/src/Twig/AssetsExtension.php
```diff
diff --git a/src/Twig/AssetsExtension.php b/src/Twig/AssetsExtension.php
index 670e5b5..d54b18e 100644
--- a/src/Twig/AssetsExtension.php
+++ b/src/Twig/AssetsExtension.php
@@ -23,6 +23,6 @@ class AssetsExtension extends AbstractExtension
 
     public function asset($path): string
     {
-        return sprintf('assets/%s', $path);
+        return sprintf('/assets/%s', $path);
     }
 }
```

https://github.com/symfony/symfony-docs/blob/d5c00035b6ca42c362cec14fcb6830555f1ac9ce/_build/build.php
```diff
diff --git a/_build/build.php b/_build/build.php
index 897fd8dac..c51904213 100755
--- a/_build/build.php
+++ b/_build/build.php
@@ -24,7 +24,7 @@ use SymfonyDocsBuilder\DocBuilder;
             ->setContentDir(__DIR__.'/..')
             ->setOutputDir($outputDir)
             ->setImagesDir(__DIR__.'/output/_images')
-            ->setImagesPublicPrefix('_images')
+            ->setImagesPublicPrefix('/_images')
             ->setTheme('rtd')
         ;
 
@@ -46,21 +46,6 @@ use SymfonyDocsBuilder\DocBuilder;
         $result = (new DocBuilder())->build($buildConfig);
 
         if ($result->isSuccessful()) {
-            // fix assets URLs to make them absolute (otherwise, they don't work in subdirectories)
-            $iterator = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($outputDir));
-
-            foreach (new RegexIterator($iterator, '/^.+\.html$/i', RegexIterator::GET_MATCH) as $match) {
-                $htmlFilePath = array_shift($match);
-                $htmlContents = file_get_contents($htmlFilePath);
-                file_put_contents($htmlFilePath, str_replace('<head>', '<head><base href="/">', $htmlContents));
-            }
-
-            foreach (new RegexIterator($iterator, '/^.+\.css/i', RegexIterator::GET_MATCH) as $match) {
-                $htmlFilePath = array_shift($match);
-                $htmlContents = file_get_contents($htmlFilePath);
-                file_put_contents($htmlFilePath, str_replace('fonts/', '../fonts/', $htmlContents));
-            }
-
             $io->success(sprintf("The Symfony Docs were successfully built at %s", realpath($outputDir)));
         } else {
             $io->error(sprintf("There were some errors while building the docs:\n\n%s\n", $result->getErrorTrace()));
```

**Comment by mdynnl** [comment]
https://github.com/symfony-tools/docs-builder/blob/faf2c5f9cd22dd00c8ef9614d521ea1cdff43390/src/Twig/AssetsExtension.php
```
diff --git a/src/Twig/AssetsExtension.php b/src/Twig/AssetsExtension.php
index 670e5b5..d54b18e 100644
--- a/src/Twig/AssetsExtension.php
+++ b/src/Twig/AssetsExtension.php
@@ -23,6 +23,6 @@ class AssetsExtension extends AbstractExtension
 
     public function asset($path): string
     {
-        return sprintf('assets/%s', $path);
+        return sprintf('/assets/%s', $path);
     }
 }
```

https://github.com/symfony/symfony-docs/blob/1179b159139c354cb27da92201d18c09b95386cd/_build/build.php
```
diff --git a/_build/build.php b/_build/build.php
index 897fd8dac..c51904213 100755
--- a/_build/build.php
+++ b/_build/build.php
@@ -24,7 +24,7 @@ use SymfonyDocsBuilder\DocBuilder;
             ->setContentDir(__DIR__.'/..')
             ->setOutputDir($outputDir)
             ->setImagesDir(__DIR__.'/output/_images')
-            ->setImagesPublicPrefix('_images')
+            ->setImagesPublicPrefix('/_images')
             ->setTheme('rtd')
         ;
 
@@ -46,21 +46,6 @@ use SymfonyDocsBuilder\DocBuilder;
         $result = (new DocBuilder())->build($buildConfig);
 
         if ($result->isSuccessful()) {
-            // fix assets URLs to make them absolute (otherwise, they don't work in subdirectories)
-            $iterator = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($outputDir));
-
-            foreach (new RegexIterator($iterator, '/^.+\.html$/i', RegexIterator::GET_MATCH) as $match) {
-                $htmlFilePath = array_shift($match);
-                $htmlContents = file_get_contents($htmlFilePath);
-                file_put_contents($htmlFilePath, str_replace('<head>', '<head><base href="/">', $htmlContents));
-            }
-
-            foreach (new RegexIterator($iterator, '/^.+\.css/i', RegexIterator::GET_MATCH) as $match) {
-                $htmlFilePath = array_shift($match);
-                $htmlContents = file_get_contents($htmlFilePath);
-                file_put_contents($htmlFilePath, str_replace('fonts/', '../fonts/', $htmlContents));
-            }
-
             $io->success(sprintf("The Symfony Docs were successfully built at %s", realpath($outputDir)));
         } else {
             $io->error(sprintf("There were some errors while building the docs:\n\n%s\n", $result->getErrorTrace()));
```

for fonts issue, I think it's better if the `fonts` is moved inside `css` than regex replace if that's allowed.

**Comment by mdynnl** [comment]
https://github.com/symfony-tools/docs-builder/blob/faf2c5f9cd22dd00c8ef9614d521ea1cdff43390/src/Twig/AssetsExtension.php
```diff
diff --git a/src/Twig/AssetsExtension.php b/src/Twig/AssetsExtension.php
index 670e5b5..d54b18e 100644
--- a/src/Twig/AssetsExtension.php
+++ b/src/Twig/AssetsExtension.php
@@ -23,6 +23,6 @@ class AssetsExtension extends AbstractExtension
 
     public function asset($path): string
     {
-        return sprintf('assets/%s', $path);
+        return sprintf('/assets/%s', $path);
     }
 }
```

https://github.com/symfony/symfony-docs/blob/1179b159139c354cb27da92201d18c09b95386cd/_build/build.php
```diff
diff --git a/_build/build.php b/_build/build.php
index 897fd8dac..c51904213 100755
--- a/_build/build.php
+++ b/_build/build.php
@@ -24,7 +24,7 @@ use SymfonyDocsBuilder\DocBuilder;
             ->setContentDir(__DIR__.'/..')
             ->setOutputDir($outputDir)
             ->setImagesDir(__DIR__.'/output/_images')
-            ->setImagesPublicPrefix('_images')
+            ->setImagesPublicPrefix('/_images')
             ->setTheme('rtd')
         ;
 
@@ -46,21 +46,6 @@ use SymfonyDocsBuilder\DocBuilder;
         $result = (new DocBuilder())->build($buildConfig);
 
         if ($result->isSuccessful()) {
-            // fix assets URLs to make them absolute (otherwise, they don't work in subdirectories)
-            $iterator = new RecursiveIteratorIterator(new RecursiveDirectoryIterator($outputDir));
-
-            foreach (new RegexIterator($iterator, '/^.+\.html$/i', RegexIterator::GET_MATCH) as $match) {
-                $htmlFilePath = array_shift($match);
-                $htmlContents = file_get_contents($htmlFilePath);
-                file_put_contents($htmlFilePath, str_replace('<head>', '<head><base href="/">', $htmlContents));
-            }
-
-            foreach (new RegexIterator($iterator, '/^.+\.css/i', RegexIterator::GET_MATCH) as $match) {
-                $htmlFilePath = array_shift($match);
-                $htmlContents = file_get_contents($htmlFilePath);
-                file_put_contents($htmlFilePath, str_replace('fonts/', '../fonts/', $htmlContents));
-            }
-
             $io->success(sprintf("The Symfony Docs were successfully built at %s", realpath($outputDir)));
         } else {
             $io->error(sprintf("There were some errors while building the docs:\n\n%s\n", $result->getErrorTrace()));
```

for fonts issue, I think it's better if the `fonts` is moved inside `css` than regex replace if that's allowed.

**Comment by javiereguiluz** [comment]
I spent some time trying to fix this and in #18844 I'm proposing a solution that doesn't require to change the `symfony-tools/docs-builder` project.

**Comment by javiereguiluz** [comment]
Closing as fixed in #18844.
